sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -65,104 +255,393 @@ public void AddLink(string id, IDictionary<string, string>? attributes = null)
+
+                if (!_currentActivity.TryDispose())
+                {
+                    _currentActivity.Stop();
+                }
             }
         }
+    }
+
+#pragma warning disable SA1507 // File can not contain multiple types
+    /// <summary>
+    /// Until we can reference the 5.0 of System.Diagnostics.DiagnosticSource
+    /// </summary>
+    internal static class ActivityExtensions
+    {
+        static ActivityExtensions()
+        {
+            ResetFeatureSwitch();
+        }
 
-        public void SetStartTime(DateTime dateTime) => _activity?.SetStartTime(dateTime);
+        private static bool SupportsActivitySourceSwitch;
 
-        public void Dispose()
+        private static readonly Type? ActivitySourceType = Type.GetType("System.Diagnostics.ActivitySource, System.Diagnostics.DiagnosticSource");
+        private static readonly Type? ActivityKindType = Type.GetType("System.Diagnostics.ActivityKind, System.Diagnostics.DiagnosticSource");
+        private static readonly Type? ActivityTagsCollectionType = Type.GetType("System.Diagnostics.ActivityTagsCollection, System.Diagnostics.DiagnosticSource");
+        private static readonly Type? ActivityLinkType = Type.GetType("System.Diagnostics.ActivityLink, System.Diagnostics.DiagnosticSource");
+        private static readonly Type? ActivityContextType = Type.GetType("System.Diagnostics.ActivityContext, System.Diagnostics.DiagnosticSource");
+
+        private static Action<Activity, int>? SetIdFormatMethod;
+        private static Func<Activity, string?>? GetTraceStateStringMethod;
+        private static Func<Activity, int>? GetIdFormatMethod;
+        private static Action<Activity, string, object?>? ActivityAddTagMethod;
+        private static Func<object, string, int, ICollection<KeyValuePair<string, object>>?, IList?, DateTimeOffset, Activity?>? ActivitySourceStartActivityMethod;
+        private static Func<object, bool>? ActivitySourceHasListenersMethod;
+        private static Func<string, ICollection<KeyValuePair<string, object>>?, object?>? CreateActivityLinkMethod;
+        private static Func<ICollection<KeyValuePair<string,object>>?>? CreateTagsCollectionMethod;
+
+        private static readonly ParameterExpression ActivityParameter = Expression.Parameter(typeof(Activity));
+
+        public static void SetW3CFormat(this Activity activity)
         {
-            if (_activity == null)
+            if (SetIdFormatMethod == null)
             {
-                return;
+                var method = typeof(Activity).GetMethod("SetIdFormat");
+                if (method == null)
+                {
+                    SetIdFormatMethod = (_, _) => { };
+                }
+                else
+                {
+                    var idParameter = Expression.Parameter(typeof(int));
+                    var convertedId = Expression.Convert(idParameter, method.GetParameters()[0].ParameterType);
```

Comment:
Won't this pull larger assemblies into the process? Can we not just do it through reflection. IIRC, though I may not, you can pass an int via reflection for a method that takes an enum, which is what it looks like you're trying to do here.

Reply:
We want logging/tracing to be as cheap as possible. Every DiagnosticScope would result in multiple reflection-based calls and those are going add up. I concluded that it's better to pay a one-time cost of precompiling these methods.